[CZBANK-3] fix: enhance number validation in ApiTransactionCreateSchema

vojtech-cerveny opened the pull request:
- Updated the preprocessing function to only accept valid number strings (integer or decimal).
- Added regex validation to ensure strings are properly formatted before parsing.
- Invalid strings will now return undefined, improving overall validation accuracy.

vojtech-cerveny committed:
fix: enhance number validation in ApiTransactionCreateSchema

- Updated the preprocessing function to only accept valid number strings (integer or decimal).
- Added regex validation to ensure strings are properly formatted before parsing.
- Invalid strings will now return undefined, improving overall validation accuracy.

vojtech-cerveny committed:
test: add unit tests for ApiTransactionCreateSchema validation

- Introduced comprehensive unit tests for ApiTransactionCreateSchema to validate both valid and invalid transaction inputs.
- Tests cover various scenarios including acceptable number formats, invalid strings, and missing fields to ensure robust schema validation.

vojtech-cerveny committed:
test: update unit tests for ApiTransactionCreateSchema to use consistent bank number format

- Modified unit tests to replace the bank number with a consistent test value for better clarity and maintainability.
- Added error logging for failed validation cases to assist in debugging and ensure comprehensive test coverage.

vojtech-cerveny committed:
fix: refine number validation in ApiTransactionCreateSchema

- Updated regex to disallow scientific notation in number validation, ensuring only valid integers and decimals are accepted.
- Adjusted unit tests to reflect this change, confirming that invalid scientific notation now results in a failed validation.

vojtech-cerveny committed:
refactor: simplify amount validation in ApiTransactionCreateSchema

- Replaced the preprocessing function with direct number validation, ensuring amounts are positive and within specified limits.
- Added transformation to round amounts to one decimal place for consistency.
- Updated unit tests to validate new constraints, including acceptance of 0.001 and rejection of amounts below this threshold.

vojtech-cerveny committed:
feat: add TODO for transaction schema improvement

- Introduced a TODO comment in transation-schema.ts to highlight the need for separate schemas for transaction creation logic and API request body.

vojtech-cerveny committed:
docs: update API best practices section

vojtech-cerveny committed:
refactor: fix name of file - transation -> transaction-schema.ts

vojtech-cerveny committed:
refactor: implement roundAmount utility for consistent amount rounding

- Introduced a new utility function `roundAmount` to safely round numbers to a specified number of decimal places, addressing floating-point precision issues.
- Updated transaction schemas to utilize `roundAmount` for rounding amounts to three decimal places.
- Added unit tests for `roundAmount` to ensure correct functionality across various scenarios, including edge cases.

vojtech-cerveny committed:
refactor: enhance transaction creation logic with improved JSON validation and amount handling

- Introduced a `validateJSON` function to handle JSON parsing errors gracefully in the transaction creation route.
- Updated the transaction transfer form to use a refined amount validation schema, ensuring amounts are processed correctly and user-friendly.
- Enhanced the amount input handling to replace commas with dots for normalization and improved user experience.
- Adjusted the button state to reflect form submission status, providing better feedback to users during transactions.

vojtech-cerveny committed:
refactor: implement withValidatedJSON for improved JSON handling in API routes

- Introduced a `withValidatedJSON` wrapper function to streamline JSON parsing and error handling across multiple API routes.
- Updated the POST handlers in bank account, transaction, and user creation routes to utilize the new validation wrapper, enhancing code consistency and reducing duplication.
- Removed redundant JSON parsing logic from individual route handlers, improving readability and maintainability.

vojtech-cerveny committed:
fix: update balance display precision in BankAccountCard and BankAccountPage

- Changed the balance display from one decimal place to three decimal places in both the BankAccountCard and BankAccountPage components for improved accuracy.

postulkam commented on src/domain/transaction-domain/transaction-schema.ts:
```diff
@@ -0,0 +1,33 @@
+export const CreateTransactionNumberToNumberSchema = z.object({
+  amount: AmountSchema,
+  currency: z.custom<Currency>(),
+  toBankNumber: z.string().endsWith("5555").length(17, "Bank number must be exactly in format 1111222233334444/5555"),
```
Can we define these `bankNumber` validations in one place? I found at least tree places with the same code in this PR - it will be easier to change in the future, if we define it once.

vojtech-cerveny replied:
Yeah, definitely this is something from my "bastleni" :) I will update it.

postulkam commented on src/domain/transaction-domain/transaction-schema.ts:
```diff
@@ -0,0 +1,33 @@
+  currency: z.custom<Currency>(),
+  toBankNumber: z.string().endsWith("5555").length(17, "Bank number must be exactly in format 1111222233334444/5555"),
+  userId: z.string(),
+  fromBankNumber: z.string(),
```
Is this ok, we don't want to validate `fromBankNumber` the same as `toBankNumber`?

vojtech-cerveny replied:
Well, for now, it is always from the application, so user doesn't have access to this field. But yeah, it would be better to make sure, that it is also checked.

postulkam reviewed (dismissed):
Two questions about validation, otherwise looks good to me and you can merge it. :)

postulkam reviewed (dismissed):
LGTM

vojtech-cerveny committed:
refactor: update transaction schemas for improved validation

- Replaced the existing bank number validation logic with a reusable `BankNumberSchema` for consistency across the application.
- Introduced `AmountWithBalanceSchema` to validate amounts against the user's balance, enhancing the transfer form's validation logic.
- Updated the `TransactionTransfer` component to utilize the new schemas for better maintainability and clarity.

postulkam reviewed (dismissed):
LGTM

postulkam reviewed (dismissed):
LGTM

vojtech-cerveny committed:
refactor: enhance error handling in API routes

- Updated the `handleErrors` function to accept both `ApiError` and `APIError`, improving flexibility in error management.
- Integrated the `handleErrors` function into the user creation route to streamline error responses and maintain consistency across API error handling.

vojtech-cerveny committed:
fix: fixing transfer issues

vojtech-cerveny commented:
@postulkam sorry for the mess - I rebased develop and I did something wrong, so commits looks like shit right now (looks like it have some newer date).. 

I tried to resolve merge conflicts (in IDE) and I will not do it again :D I can delete this mess and start again, let me know. Or LGTM and we can rid of this mess :D

postulkam commented:
Did you sometimes merged develop and other times rebased onto it? This could case your problem. You have to pick one approach (merge x rebase) and stick with it for the whole PR lifetime.

> @postulkam sorry for the mess - I rebased develop and I did something wrong, so commits looks like shit right now (looks like it have some newer date)..
> 
> I tried to resolve merge conflicts (in IDE) and I will not do it again :D I can delete this mess and start again, let me know. Or LGTM and we can rid of this mess :D

This is up to you, do you want to use the code? 

If yes, there is another "way" to "merge" it into develop.
Crate new branch from develop and cherry-pick the commits one by one. If I get lost, I usually do it like that - you will solve problems with each commit separately, and since you’re doing it in small chunks, the margin of error is smaller.

vojtech-cerveny commented:
No, when we are squashing, I don't care about commits in PR. I fixed it by reverting stuff, so code is the same what I want, but commit history is the mess. And yeah, I will stick with it forever :) 

So for me it is ready to review and ready to merge

postulkam commented:
@vojtech-cerveny can you please merge develop here (I think, you will have so many conflicts) and I will make review after? :)

vojtech-cerveny commented:
Sure, will do!

vojtech-cerveny commented:
Will move it into new PR.
